probe_src/python/probe_py/manual/transformations.py

```diff
@@ -0,0 +1,181 @@
+import argparse
+
+import networkx as nx
+import pathlib
+import dataclasses
+import os
+from typing import Tuple, List
+
+
+## Assuming this is definition of the dataflow graph
+@dataclasses.dataclass(frozen=True)
+class FileNode:
+    inode: int
+    version: int
+    fileName: str = None
+
+@dataclasses.dataclass(frozen=True)
+class ProcessNode:
+    cmd: Tuple[str, ...]
+
+A = FileNode(0, 0, "A.txt")
+B0 = FileNode(1, 0, "B.txt")
+B1 = FileNode(1, 1, "B.txt")
+C = FileNode(3, 0, "C.txt")
+W = ProcessNode(("cp", "A.txt", "B.txt"))
+X = ProcessNode(("sed", "s/foo/bar/g", "-i", "B.txt"))
+# Note, the filename in FileNode will not always appear in the cmd of ProcessNode!
+Y = ProcessNode(("analyze", "-i", "-k"))
+
+
+example_dataflow_graph = nx.DiGraph()
+# FileNodes will be red and ProcessNodes will be blue in the visualization
+# Code can distinguish between the two using isinstance(node, ProcessNode) or likewise with FileNode
+example_dataflow_graph.add_nodes_from([A, B0, B1, C], color="red")
+example_dataflow_graph.add_nodes_from([W, X, Y], color="blue")
+example_dataflow_graph.add_edges_from([
+    (A, W),
+    (W, B0),
+    (B0, X),
+    (X, B1),
+    (A, Y),
+    (B1, Y),
+    (Y, C),
+])
+
+# Function to generate a unique process name
+def generate_process_name(base_name, process_counter):
+    count = process_counter.get(base_name, 0)
+    process_name = f"{base_name}_{count}"
+    process_counter[base_name] = count + 1
+    return process_name
+
+def escape_filename_for_nextflow(filename: str) -> str:
+    """
+    Escape special characters in a filename for Nextflow.
+    Replace any non-letter, non-number character with _i, where i is the ASCII character code in hex.
+    Additionally, escape underscores and prepend an escape code if the filename starts with a number.
+    """
+    escaped_filename = []
+
+    for char in filename:
+        if char.isalnum():  # Keep letters and numbers unchanged
+            escaped_filename.append(char)
+        elif char == '_':  # Escape underscores
+            continue
```

Comment:
I think this will erase the underscores. Rather than `continue`, just append `_{ord(char):02x}`